libcontainer/capabilities/capabilities.go

```diff
@@ -49,19 +52,55 @@ func KnownCapabilities() []string {
 // or Capabilities that are unavailable in the current environment are ignored,
 // printing a warning instead.
 func New(capConfig *configs.Capabilities) (*Caps, error) {
-	var (
-		err error
-		c   Caps
-	)
+	var c Caps
 
+	cm, err := capMap()
+	if err != nil {
+		return nil, err
+	}
 	unknownCaps := make(map[string]struct{})
+	ignoredCaps := make(map[string]struct{})
+	// capSlice converts the slice of capability names in caps, to their numeric
+	// equivalent, and returns them as a slice. Unknown or unavailable capabilities
+	// are not returned, but appended to unknownCaps.
+	//
+	// If mustHave argument is not nil, caps that are not present in mustHave
+	// are appended to ignoredCaps instead of the resulting slice.
+	capSlice := func(caps []string, mustHave []capability.Cap) []capability.Cap {
+		out := make([]capability.Cap, 0, len(caps))
+		for _, c := range caps {
+			if v, ok := cm[c]; !ok {
+				unknownCaps[c] = struct{}{}
+			} else {
+				if mustHave != nil && !slices.Contains(mustHave, v) {
+					ignoredCaps[c] = struct{}{}
+					continue
+				}
+				out = append(out, v)
+			}
+		}
+		return out
+	}
+	inheritable := capSlice(capConfig.Inheritable, nil)
+	// Ambient vector can not contain values not raised in the Inheritable vector,
+	// and errors setting ambient capabilities were previously ignored due to a bug
+	// (see https://github.com/kolyshkin/capability/pull/3), so to maintain backward
+	// compatibility we have to ignore those Ambient caps that are not also raised
+	// in Inheritable (and issue a warning).
+	ambient := capSlice(capConfig.Ambient, inheritable)
+	if len(ignoredCaps) > 0 {
```

Comment:
This PR handles condition 1. As for condition 2, it works as it used to work before, no changes needed here.

In general, I think, we should return an error when we're unable to set a capability requested. The warning which this PR adds is a temporary measure, to ensure we're not breaking our users. Eventually this warning should become an error.

Reply:
> This PR handles condition 1.

Because of ‘both permitted and inheritable cap sets’.

Even in condition 1, I think you need to add check ‘permitted’ too, not only ‘inheritable’. Please see my first commit in #4418 .



> As for condition 2, it works as it used to work before, no changes needed here.

I think this error was also masked before this PR, but will be failed in the future. It should have a compatibility like condition 1.

Reply:
> Even in condition 1, I think you need to add check ‘permitted’ too, not only ‘inheritable’. Please see my first commit in #4418 .

It looks like such configuration is not working in the current runc version:

```
@test "runc run [raise inheritable bit not set in permitted]" {
        update_config '   .process.capabilities.inheritable = ["CAP_KILL"]
                        | .process.capabilities.permitted = ["CAP_CHOWN"]'
#                       | .process.capabilities.ambient = ["CAP_KILL", "CAP_CHOWN"]'

        runc run testct
        [ "$status" -eq 0 ]
```

```
 runc run testct (status=1):
   time="2024-09-27T17:31:28-07:00" level=error msg="runc run failed: unable to start container process: error during container init: unable to apply caps: operation not permitted"
```

This means:
1. If it's not working now, it should not start work in the future. 
2. We don't have to check for permitted bit set when checking ambient.

> > As for condition 2, it works as it used to work before, no changes needed here.

> I think this error was also masked before this PR, but will be failed in the future. It should have a compatibility like condition 1.

You might be right, I need to check it.